WIP: Fix missing borders in tables with border-collapse

mbrubeck opened the pull request:
This fixes a pair of bugs that caused some borders to be ignored by the border-collapse code. r? @notriddle 

I haven't done a full test run with this change yet. After a `try` build I will update test expectations and/or add a new test.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #14834 (github issue number if applicable).

- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/15156)
<!-- Reviewable:end -->

mbrubeck committed:
Fix provenance of first cell's inline collapsed border

This fixes the inline-direction part of #14834.

mbrubeck committed:
Fix block-direction border collapsing calculation

Previously the block-start collapsed borders from rows/cells were
written to `preliminary_collapsed_borders` but never read.  This fixes
the block-direction part of #14834.

highfive commented:
Heads up! This PR modifies the following files:
 * @emilio: components/layout/table.rs, components/layout/table_row.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify layout code, but no tests are modified. Please consider adding a test!

mbrubeck commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 33ac138 with merge 3ae18b2...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/5015), [arm32](http://build.servo.org/builders/arm32/builds/5185), [arm64](http://build.servo.org/builders/arm64/builds/5229), [linux-dev](http://build.servo.org/builders/linux-dev/builds/5331), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/2005), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/2016), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/5302), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/5342), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/2003), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/2050), [windows-gnu-dev](http://build.servo.org/builders/windows-gnu-dev/builds/469), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/474)

mbrubeck commented:
Updated test expectations and fixed a bug.

The try run above was affected by #15158, so it incorrectly showed green tests.  This should run through try again after that fix lands.

mbrubeck commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 18d271b with merge 2385064...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/2031)

notriddle commented:
[Some of the tests you enabled failed on Mac](http://build.servo.org/builders/mac-rel-css/builds/5365/steps/shell__1/logs/filtered-css-errorsummary.log), while the one that failed on Linux passed on Mac. Fonts?

mbrubeck commented:
Updated test expectations again to match the Try results. (The previous update was based on a local test run; my local fonts must differ from the Linux builders.)

notriddle commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 258162f has been approved by `notriddle`

<!-- @bors-servo r=notriddle 258162fbee65fb894b8aac5dbce5a366f71dc7b3 -->

bors-servo commented:
:hourglass: Testing commit 258162f with merge 1b15578...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/5038), [arm32](http://build.servo.org/builders/arm32/builds/5215), [arm64](http://build.servo.org/builders/arm64/builds/5255), [linux-dev](http://build.servo.org/builders/linux-dev/builds/5359), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/2037), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/2047), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/5331), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/5371), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/2030), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/2082), [windows-gnu-dev](http://build.servo.org/builders/windows-gnu-dev/builds/497), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/504)
